### PR TITLE
add unet

### PR DESCRIPTION
**Pull request by Junjun2016**
add unet

### CODE REVIEW

**Comment by hellock** [comment]
Task linked: [CU-4fp4hc UNet 模型](https://app.clickup.com/t/4fp4hc)

File: mmseg/models/backbones/unet.py [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+            Default: dict(type='BN').
+        act_cfg (dict | None): Config dict for activation layer in ConvModule.
+            Default: dict(type='ReLU').
+        upsample (bool): Whether need upsample the high-level feature map
+            or not. If the size of high-level feature map is the same as that
+            of skip feature map (low-level feature map from encoder), it does
+            not need upsample the high-level feature map and the upsample
+            is False. Default: True.
+        upsample_cfg (dict): The upsample config of the upsample module in
+            decoder. Default: dict(type='interp_up').
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:257]: I suggest merging `upsample` with `upsample_cfg`. When `upsample_cfg is None`, there will be no upsample.

File: mmseg/models/backbones/unet.py [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+                 with_cp=False,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 *,
+                 kernel_size=4,
+                 stride=2,
+                 padding=1):
+        super(DeconvUpsample, self).__init__()
+
+        assert stride == 2 and (kernel_size-2*padding-2) == 0, \
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:120]: We may calculate padding directly.

File: mmseg/models/backbones/unet.py [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+class BasicConvBlock(nn.Module):
+    """Basic convolutional block for UNet.
+
+    This module consists of several plain convolutional layers.
+
+    Args:
+        in_channels (int): Number of input channels.
+        out_channels (int): Number of output channels.
+        num_convs (int): Number of convolutional layers. Default: 2.
+        stride (int 1|2): Whether use stride convolution to downsample
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:23]:
```suggestion
        stride (int): Whether use stride convolution to downsample
```

File: mmseg/models/backbones/unet.py [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+            the input feature map. If stride=2, it only uses stride convolution
+            in the first convolutional layer to downsample the input feature
+            map. Default: 1.
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:26]:
```suggestion
            map. Options are 1 or 2. Default: 1.
```

File: mmseg/models/backbones/unet.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+                 *,
+                 kernel_size=4):
+        super(DeconvUpsample, self).__init__()
+
+        assert (kernel_size - 2 >= 0) and (kernel_size - 2) % 2 == 0,\
+            f'kernel_size should be even numbers and greater than or equal to \
+            2, while the kernel size is {kernel_size}.'
+
+        stride = 2
+        padding = (kernel_size - 2) // 2
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:120]: There will be errors (unittest should cover this).

File: mmseg/models/backbones/unet.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+            in_channels = base_channels * 2**i
+
+    def forward(self, x):
+        self._check_input_devisible(x)
+        enc_outs = []
+        for enc in self.encoder:
+            x = enc(x)
+            enc_outs.append(x)
+        dec_outs = [x]
+        for i in range(len(self.decoder) - 1, -1, -1):
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:378]: We may use `reverse`.

File: mmseg/models/backbones/unet.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+                 out_channels,
+                 num_convs=2,
+                 stride=1,
+                 dilation=1,
+                 with_cp=False,
+                 conv_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 act_cfg=dict(type='ReLU'),
+                 dcn=None,
+                 plugins=None):
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:54]:
```suggestion
                 plugins=None): plugins=None
```

**Review by xvjiarui — APPROVED** [review]

File: mmseg/models/backbones/unet.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        """Forward function."""
+
+        if self.with_cp and x.requires_grad:
+            out = cp.checkpoint(self.convs, x)
+        else:
+            out = self.convs(x)
+        return out
+
+
+@UPSAMPLE_LAYERS.register_module(name='deconv_up2x')
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:86]: We may support 2x, 4x.

File: mmseg/models/backbones/unet.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        """Forward function."""
+
+        if self.with_cp and x.requires_grad:
+            out = cp.checkpoint(self.deconv_upsamping, x)
+        else:
+            out = self.deconv_upsamping(x)
+        return out
+
+
+@UPSAMPLE_LAYERS.register_module(name='interp_up')
```
Comment (xvjiarui) [inline mmseg/models/backbones/unet.py:143]:
```suggestion
@UPSAMPLE_LAYERS.register_module(name='interp_conv')
```

**Comment by OpenMMLab-Assistant003** [comment]
Hi @Junjun2016！First of all, we want to express our gratitude for your significant PR in the MMSegmentation project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤ @Junjun2016
